File: implementation/scipamato/core/web/src/test/java/ch/difty/scipamato/core/web/paper/jasper/CoreShortFieldWithEmptyMainFieldConcatenatorTest.java
package ch.difty.scipamato.core.web.paper.jasper;

import static ch.difty.scipamato.common.TestUtilsKt.assertDegenerateSupplierParameter;
import static org.assertj.core.api.AssertionsForClassTypes.assertThat;

import org.junit.jupiter.api.Test;

import ch.difty.scipamato.core.entity.Paper;

@SuppressWarnings("SpellCheckingInspection")
class CoreShortFieldWithEmptyMainFieldConcatenatorTest {

    private final CoreShortFieldConcatenator sfc = new CoreShortFieldWithEmptyMainFieldConcatenator();

    private final Paper p = new Paper();

    @SuppressWarnings("SpellCheckingInspection")
    private final ReportHeaderFields rhf = ReportHeaderFields
        .builder("hp", "b")
        .methodStudyDesignLabel("msdl")
        .methodOutcomeLabel("mol")
        .populationPlaceLabel("ppll")
        .exposurePollutantLabel("epl")
        .exposureAssessmentLabel("eal")
        .methodStatisticsLabel("msl")
        .methodConfoundersLabel("mcl")
        .populationDurationLabel("pdl")
        .populationParticipantsLabel("ppal")
        .resultMeasuredOutcomeLabel("rmol")
        .resultEffectEstimateLabel("reel")
        .resultExposureRangeLabel("rerl")
        .conclusionLabel("ccl")
        .build();

    @Test
    void methods_withNullRecordset_throws() {
        assertDegenerateSupplierParameter(() -> sfc.methodsFrom(null, rhf), "paper");
    }

    @Test
    void population_withNullRecordset_throws() {
        assertDegenerateSupplierParameter(() -> sfc.populationFrom(null, rhf), "paper");
    }

    @Test
    void result_withNullRecordset_throws() {
        assertDegenerateSupplierParameter(() -> sfc.resultFrom(null, rhf), "paper");
    }

    @Test
    void methods_withNonNullMethod_returnsMethod() {
        p.setMethods("method");
        p.setMethodStudyDesign("msd");
        p.setMethodOutcome("mo");
        p.setPopulationPlace("pp");
        p.setExposurePollutant("ep");
        p.setExposureAssessment("ea");
        p.setMethodStatistics("ms");
        p.setMethodConfounders("mc");

        assertThat(sfc.methodsFrom(p, rhf)).isEqualTo("method");
    }

    @Test
    void methods_withNullMethod_returnsConcatenatedShortMethodFieldsConcatenated() {
        p.setMethods(null);
        p.setMethodStudyDesign("msd");
        p.setMethodOutcome("mo");
        p.setPopulationPlace("pp");
        p.setExposurePollutant("ep");
        p.setExposureAssessment("ea");
        p.setMethodStatistics("ms");
        p.setMethodConfounders("mc");

        assertThat(sfc.methodsFrom(p, rhf)).isEqualTo(
            "msdl: msd\nmol: mo\nppll: pp\nepl: ep\neal: ea\nmsl: ms\nmcl: mc");
    }

    @Test
    void population_withNonNullPopulation_returnsPopulation() {
        p.setPopulation("population");
        p.setPopulationPlace("ppl");
        p.setPopulationParticipants("ppa");
        p.setPopulationDuration("pd");

        assertThat(sfc.populationFrom(p, rhf)).isEqualTo("population");
    }

    @Test
    void population_withNullPopulation_returnsPopulationShortFieldsConcatenated() {
        p.setPopulation(null);
        p.setPopulationPlace("ppl");
        p.setPopulationParticipants("ppa");
        p.setPopulationDuration("pd");

        assertThat(sfc.populationFrom(p, rhf)).isEqualTo("ppll: ppl\nppal: ppa\npdl: pd");
    }

    @Test
    void result_withNonNullResult_returnsResult() {
        p.setResult("result");
        p.setResultMeasuredOutcome("rmo");
        p.setResultExposureRange("rer");
        p.setResultEffectEstimate("ree");
        p.setConclusion("cc");

        assertThat(sfc.resultFrom(p, rhf)).isEqualTo("result");
    }

    @Test
    void result_withNullResult_returnsResultShortFieldsConcatenated() {
        p.setResult(null);
        p.setResultMeasuredOutcome("rmo");
        p.setResultExposureRange("rer");
        p.setResultEffectEstimate("ree");
        p.setConclusion("cc");

        assertThat(sfc.resultFrom(p, rhf)).isEqualTo("rmol: rmo\nrerl: rer\nreel: ree\nccl: cc");
    }

}

File: core/core-web/src/test/kotlin/ch/difty/scipamato/core/web/paper/jasper/CoreShortFieldWithEmptyMainFieldConcatenatorTest.kt
package ch.difty.scipamato.core.web.paper.jasper

import ch.difty.scipamato.core.entity.Paper
import org.amshove.kluent.shouldBeEqualTo
import org.junit.jupiter.api.Test

internal class CoreShortFieldWithEmptyMainFieldConcatenatorTest {

    private val sfc: CoreShortFieldConcatenator = CoreShortFieldWithEmptyMainFieldConcatenator()
    private val p = Paper()
    private val rhf = ReportHeaderFields
        .builder("hp", "b")
        .methodStudyDesignLabel("msdl")
        .methodOutcomeLabel("mol")
        .populationPlaceLabel("ppll")
        .exposurePollutantLabel("epl")
        .exposureAssessmentLabel("eal")
        .methodStatisticsLabel("msl")
        .methodConfoundersLabel("mcl")
        .populationDurationLabel("pdl")
        .populationParticipantsLabel("ppal")
        .resultMeasuredOutcomeLabel("rmol")
        .resultEffectEstimateLabel("reel")
        .resultExposureRangeLabel("rerl")
        .conclusionLabel("ccl")
        .build()

    @Test
    fun methods_withNonNullMethod_returnsMethod() {
        p.methods = "method"
        p.methodStudyDesign = "msd"
        p.methodOutcome = "mo"
        p.populationPlace = "pp"
        p.exposurePollutant = "ep"
        p.exposureAssessment = "ea"
        p.methodStatistics = "ms"
        p.methodConfounders = "mc"
        sfc.methodsFrom(p, rhf) shouldBeEqualTo "method"
    }

    @Test
    fun methods_withNullMethod_returnsConcatenatedShortMethodFieldsConcatenated() {
        p.methods = null
        p.methodStudyDesign = "msd"
        p.methodOutcome = "mo"
        p.populationPlace = "pp"
        p.exposurePollutant = "ep"
        p.exposureAssessment = "ea"
        p.methodStatistics = "ms"
        p.methodConfounders = "mc"
        sfc.methodsFrom(p, rhf) shouldBeEqualTo
            "msdl: msd\nmol: mo\nppll: pp\nepl: ep\neal: ea\nmsl: ms\nmcl: mc"
    }

    @Test
    fun population_withNonNullPopulation_returnsPopulation() {
        p.population = "population"
        p.populationPlace = "ppl"
        p.populationParticipants = "ppa"
        p.populationDuration = "pd"
        sfc.populationFrom(p, rhf) shouldBeEqualTo "population"
    }

    @Test
    fun population_withNullPopulation_returnsPopulationShortFieldsConcatenated() {
        p.population = null
        p.populationPlace = "ppl"
        p.populationParticipants = "ppa"
        p.populationDuration = "pd"
        sfc.populationFrom(p, rhf) shouldBeEqualTo "ppll: ppl\nppal: ppa\npdl: pd"
    }

    @Test
    fun result_withNonNullResult_returnsResult() {
        p.result = "result"
        p.resultMeasuredOutcome = "rmo"
        p.resultExposureRange = "rer"
        p.resultEffectEstimate = "ree"
        p.conclusion = "cc"
        sfc.resultFrom(p, rhf) shouldBeEqualTo "result"
    }

    @Test
    fun result_withNullResult_returnsResultShortFieldsConcatenated() {
        p.result = null
        p.resultMeasuredOutcome = "rmo"
        p.resultExposureRange = "rer"
        p.resultEffectEstimate = "ree"
        p.conclusion = "cc"
        sfc.resultFrom(p, rhf) shouldBeEqualTo "rmol: rmo\nrerl: rer\nreel: ree\nccl: cc"
    }
}
